src/webgpu/api/operation/resource_init/buffer.spec.ts

```diff
@@ -1,11 +1,231 @@
+  GPUConst.BufferUsage.COPY_SRC | GPUConst.BufferUsage.MAP_WRITE,
+  GPUConst.BufferUsage.COPY_SRC,
+];
+
+class F extends GPUTest {
+  GetBufferUsageFromMapMode(mapMode: GPUMapModeFlags): number {
+    switch (mapMode) {
+      case GPUMapMode.READ:
+        return GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ;
+      case GPUMapMode.WRITE:
+        return GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE;
+      default:
+        unreachable();
+        return 0;
+    }
+  }
+
+  async CheckGPUBufferContent(
+    buffer: GPUBuffer,
+    bufferUsage: GPUBufferUsageFlags,
+    expectedData: Uint8Array
+  ): Promise<void> {
+    // We can only check the buffer contents with t.expectContents() when the buffer usage contains
+    // COPY_SRC.
+    if (bufferUsage & GPUBufferUsage.MAP_READ) {
+      await buffer.mapAsync(GPUMapMode.READ);
+      this.expectBuffer(new Uint8Array(buffer.getMappedRange()), expectedData);
+      buffer.unmap();
+    } else {
+      assert((bufferUsage & GPUBufferUsage.COPY_SRC) !== 0);
+      this.expectContents(buffer, expectedData);
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('partial_write_buffer')
+  .desc(
+    `
+  Verify when we upload data to a part of a buffer with writeBuffer() just after the creation of the
+  buffer, the remaining part of that buffer will be initialized to 0.
+  `
+  )
+  .paramsSubcasesOnly(u => u.combine('offset', [0, 8, -12]))
+  .fn(async t => {
+    const { offset } = t.params;
+    const bufferSize = 32;
+    const appliedOffset = offset >= 0 ? offset : bufferSize + offset;
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const copySize = 12;
+    const writeData = new Uint8Array(copySize);
+    const expectedData = new Uint8Array(bufferSize);
+    for (let i = 0; i < copySize; ++i) {
+      expectedData[appliedOffset + i] = writeData[i] = i + 1;
+    }
+    t.queue.writeBuffer(buffer, appliedOffset, writeData, 0);
+
+    t.expectContents(buffer, expectedData);
+  });
+
+g.test('map_whole_buffer')
+  .desc(
+    `
+  Verify when we map the whole range of a mappable GPUBuffer to a typed array buffer just after
+  creating the GPUBuffer, the contents of both the typed array buffer and the GPUBuffer itself have
+  already been initialized to 0.
+  `
+  )
+  .params(u => u.combine('mapMode', kMapModeOptions))
+  .fn(async t => {
+    const { mapMode } = t.params;
+
+    const bufferSize = 32;
+    const bufferUsage = t.GetBufferUsageFromMapMode(mapMode);
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: bufferUsage,
+    });
+
+    await buffer.mapAsync(mapMode);
+    const readData = new Uint8Array(buffer.getMappedRange());
+    for (let i = 0; i < bufferSize; ++i) {
+      t.expect(readData[i] === 0);
+    }
+    buffer.unmap();
+
+    // We can only check the buffer contents with t.expectContents() when the buffer usage contains
```

Comment:
nit: I think this can use await t.CheckGPUBufferContent(buffer, bufferUsage, expectedData); ?